engine/src/test/java/io/zeebe/engine/processing/variable/UpdateVariableDocumentProcessorTest.java

```diff
@@ -7,146 +7,208 @@
+            .withUpdateSemantic(VariableDocumentUpdateSemantic.PROPAGATE)
+            .update();
 
     // then
-    verify(variablesState, times(1))
-        .setVariablesFromDocument(instance.getKey(), WORKFLOW_KEY, variables);
+    final Record<VariableRecordValue> createdVariable =
+        RecordingExporter.variableRecords(VariableIntent.CREATED)
+            .skipUntil(r -> r.getPosition() > result.getSourceRecordPosition())
+            .withScopeKey(workflowInstanceKey)
+            .withWorkflowInstanceKey(workflowInstanceKey)
+            .limit(document.size())
+            .findFirst()
+            .orElseThrow();
+    final Record<VariableRecordValue> updatedVariable =
+        RecordingExporter.variableRecords(VariableIntent.UPDATED)
+            .withScopeKey(workflowInstanceKey)
+            .withWorkflowInstanceKey(workflowInstanceKey)
+            .limit(document.size())
+            .findFirst()
+            .orElseThrow();
+    assertThat(result).hasRecordType(RecordType.EVENT).hasIntent(VariableDocumentIntent.UPDATED);
+    assertThat(createdVariable.getValue()).hasName("created").hasValue("1");
+    assertThat(updatedVariable.getValue()).hasName("updated").hasValue("\"newValue\"");
   }
 
   @Test
-  public void shouldApplyLocalUpdateOperation() {
+  public void shouldNotPropagateValueWithLocalSemantic() {
     // given
-    final DirectBuffer variables = asMsgPack("a", 1);
-    final ElementInstance instance = newElementInstance();
-    final TypedRecord<VariableDocumentRecord> command = newCommand(VariableDocumentRecord.class);
-    command
-        .getValue()
-        .setScopeKey(instance.getKey())
-        .setUpdateSemantics(VariableDocumentUpdateSemantic.LOCAL)
-        .setVariables(variables);
+    final Map<String, Object> document = Map.of("updated", "newValue");
+    final long workflowInstanceKey = startProcessWithVariables(Map.of("updated", "oldValue"));
+    final long serviceTaskScopeKey = getServiceTaskScopeKey();
 
     // when
-    processor.onCommand(command, controller);
+    final Record<VariableDocumentRecordValue> result =
+        ENGINE
+            .variables()
+            .ofScope(serviceTaskScopeKey)
+            .withDocument(document)
+            .withUpdateSemantic(VariableDocumentUpdateSemantic.LOCAL)
+            .update();
 
     // then
-    verify(variablesState, times(1))
-        .setVariablesLocalFromDocument(instance.getKey(), WORKFLOW_KEY, variables);
+    final Record<VariableRecordValue> createdVariable =
+        RecordingExporter.variableRecords(VariableIntent.CREATED)
+            .skipUntil(r -> r.getPosition() > result.getSourceRecordPosition())
+            .withScopeKey(serviceTaskScopeKey)
+            .withWorkflowInstanceKey(workflowInstanceKey)
+            .limit(document.size())
+            .findFirst()
+            .orElseThrow();
+    assertThat(result).hasRecordType(RecordType.EVENT).hasIntent(VariableDocumentIntent.UPDATED);
```

Comment:
This test is named `shouldNotPropagate`, but then it updates the variable document which exists at the process scope. Shouldn't it just have created a new one at the service task scope?

Reply:
So it created a variable, yes, but it didn't create a document or anything. `UPDATED` here is just the response for `UPDATE`. I can understand the confusion - what would you propose? The hard part here is, if you have many scopes and potentially modified all of them, what is the correct response representation to convey the update results as a single record?

At any rate, to know if the variable was created/updated, you can still find out by looking at the individual variable records which are also produced, which the test does first.

Reply:
Looking at this again, I think the test itself is clear enough. The UPDATED response is separate from the Variable records, I just did not understand this at the time, but that's just me not knowing about variable UPDATE responses 😄 

I wonder why testing for the absence is difficult. After checking that the variable was CREATED, we can also verify that there exists no variable UPDATED on the process scope right?

Reply:
Because the way we test for absence with `RecordingExporter` is waiting 5 seconds, and if we don't see anything, we assume that it means it's not there, which isn't necessarily true :sweat_smile:

Reply:
@npepinpe what about?
```suggestion
    assertThat(result).hasRecordType(RecordType.EVENT).hasIntent(VariableDocumentIntent.UPDATED);
    Assertions.assertThat(
            RecordingExporter.getRecords().stream()
                .filter(r -> r.getIntent() == VariableIntent.UPDATED))
        .isEmpty();
```

Reply:
This bypasses the waiting afaik

Reply:
Ah, sure, but then if you access the records directly you may be looking at them before all records have been exported since exporting is async.

Chris mentioned a better solution which is sending a second a command and waiting for that one. I think it's OK now as we don't typically do this anyway, but we could start doing that later on.